cpp/docs/TRANSITIONGUIDE.md

```diff
@@ -0,0 +1,206 @@
+```
+
+A simple wrapper class called `cudf::table` is defined in `cudf/cpp/src/table/table.hpp` to provide a convenient abstraction for the above use case. This class should be used any time a function is operating on a collection of `gdf_column`s of equal size.
+
+For example, the above becomes:
+```
+void some_multi_col_function(cudf::table & table);
+```
+
+## Input/Output Style
+
+The preferred style for how inputs are passed in and outputs are returned is the following:
+
+* Inputs 
+   * Expensive to copy types --- pass by constant reference
+   * Inexpensive to copy types --- pass by value
+* In/Outs - Pass via pointer
+* Outputs - Returned by value
+   * Returning outputs by value will not incur a copy in most cases. For more details, see: https://stackoverflow.com/questions/12953127/what-are-copy-elision-and-return-value-optimization
+   * **NOTE:** This may mean returning a `gdf_column` whose device memory was allocated within a `libcudf` function and it is the caller's responsibility to free the underlying memory. This is a temporary headache until a `cudf::column` abstraction is designed to allieviate it. 
+
+For example:
+
+```
+namespace cudf{
+column some_function( cudf::table const& input_table, cudf::table * input_output_table, int size){
+  column col;
+  // Do stuff to create output col
+  ...
+  // Return the result column of this function
+  return col;
+}
+}
+```
+
+And the usage would be:
+```
+cudf::table input;
+cudf::table in_out;
+int size;
+column output = cudf::some_function(input, &in_out, size);
+```
+
+### Multiple Return Values
+
+Sometimes it is necessary for functions to have multiple outputs. There are a few ways this can be done in C++ (including creating a `struct` for the output). One convenient way to do this is using `std::tie` and `std::make_tuple`.
+
+```
+auto return_multiple_outputs(void){
+  cudf::table out0;
+  cudf::table out1;
+  ...
+  // Do stuff with out0, out1
+  
+  // Return a tuple of all the outputs
+  return std::make_tuple(out0, out1);
+}
+
+
+cudf::table out0;
+cudf::table out1;
+std::tie(out0, out1) = cudf::return_multiple_outputs();
+```
+
+
+## Error Checking
+
+The `gdf_error` codes are deprecated. Instead, `libcudf` functions should prefer throwing exceptions.
+
+In place of the `GDF_REQUIRES` convenience macro, the `CUDF_EXPECTS` macro is provided in `cudf/cpp/src/utilities/error_utils.hpp`. 
+
+Similar to `GDF_REQUIRES`, `CUDF_EXPECTS` will throw an exception if a condition is violated. The first argument is the conditional expression that can be evaluated as a `boolean` expression and is expected to resolve to `true` under normal conditions. If the conditional evaluates to `false`, then an error has occurred and an instance of `cudf::logic_error` is thrown. The second argument to `CUDF_EXPECTS` is a short description of the error that has occurred if the conditional is false. 
+
+Example usage: 
+
+```
+CUDF_EXPECTS(lhs->dtype == rhs->dtype, "Column type mismatch");
```

Comment:
Not a good example - we have a utility function for that... well, in a PR anyway. At least mention the version of this with the utility function.